include/tscore/ts_assert.h

```diff
@@ -0,0 +1,53 @@
+/** @file
+
+  A brief file description
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include "tscore/ink_assert.h"
+
+/*
+TS_RELEASE_ASSERT expands to a void-type expression.  It causes the program to abort if an expression EXPR of
+type bool (or of a type that implictly converts to bool) is false.
+*/
+#define TS_RELEASE_ASSERT(EXPR) ((EXPR) ? static_cast<void>(0) : _ink_assert(#EXPR, __FILE__, __LINE__))
+
+/*
+TS_ASSERT/TS_ASSERT_NSE expand to void-type expressions.  They cause the program to abort, in a non-relaase or
+otherwise non-optimized executable, if an expression EXPR of type bool (or of a type that implictly converts to
+bool) is false.  For TS_ASSERT, side-effects of EXPR will still occur even in a non-optimized executable.  For
+TS_ASSERT_NSE (NSE = no side effects), side-effects of EXPR will not occur, in a non-optimized executable.
+*/
+
+// __OPTIMIZE__ is defined by both gcc and clang/LLVM if optimization level is 1 or higher.
+//
+#if defined(__OPTIMIZE__)
+
+#define TS_ASSERT(EXPR) static_cast<void>(EXPR)
+#define TS_ASSERT_NSE(EXPR) static_cast<void>(0)
```

Comment:
Why we need these two asserts? Is there any case we can't use NSE version?

Reply:
You can see why in this code:  https://godbolt.org/z/ktjtsN .  Lines 29 and 30 are equivalent.  If the tested expression doesn't call any functions, the compiler will optimize out the evaluation of the expression, to TS_ASSERT effectively is the same as TS_ASSERT_NSE.  In line 31, it is of value to use TS_ASSERT_NSE, to avoid an unnecessary function call in the release build.  In line 32, it is important to use TS_ASSERT, because we still want to call the function in the expression in the release build.

Reply:
Thanks for the description.

> In line 32, it is important to use TS_ASSERT, because we still want to call the function in the expression in the release build.

I agree this is the key point. I'm wondering we have this use case in practice.
In this case, when `TSVConnTunnel(ssl)` returns `TS_ERROR` in the release build, that error is ignored. 
It looks like the error should be handled or release assert should be used.

Reply:
To be honest, I only vaguely know what TSVConnTunnel() does.  If the only reason it would return an error is parameter validation, then a debug-only assert would likely be OK.  If there's an error case where some system resource is found to be exhausted, then it does need a release assert.  If there's an error case where the connection has been dropped, then foo() would need logic to recover from the error, no assert at all, IMO.

Reply:
There are, unfortunately, many cases where side effects are inside asserts. I think that's a terrible thing to do, but I see PRs even now that are adding code with such side effect asserts and I was brushed backed when I objected to it.